Sources/SwiftFusion/Inference/FlattenedScalars.swift

```diff
@@ -0,0 +1,154 @@
+// TODO: Move all these to Penguin.
+
+import PenguinStructures
+
+public struct FlattenedScalars<Base: MutableCollection>
+  where Base.Element: Vector
+{
+  var base: Base
+    
+  init(_ base: Base) {
+    self.base = base
+  }
+}
+
+extension FlattenedScalars: Sequence {
+  public typealias Element = Base.Element.Scalars.Element
+  
+  public struct Iterator: IteratorProtocol {
+    var outer: Base.Iterator
+    var inner: Base.Element.Scalars.Iterator?
+
+    init(outer: Base.Iterator, inner: Base.Element.Scalars.Iterator? = nil) {
+      self.outer = outer
+      self.inner = inner
+    }
+
+    public mutating func next() -> Element? {
+      while true {
+        if let r = inner?.next() { return r }
+        guard let o = outer.next() else { return nil }
+        inner = o.scalars.makeIterator()
+      }
+    }
+  }
+
+  public func makeIterator() -> Iterator { Iterator(outer: base.makeIterator()) }
+
+  // https://bugs.swift.org/browse/SR-13486 points out that the withContiguousStorageIfAvailable
+  // method is dangerously underspecified, so we don't implement it here.
+  
+  public func _customContainsEquatableElement(_ e: Element) -> Bool? {
```

Comment:
Wow, there are lots of secret stdlib protocol requirements. I don't know what they are for but I'd be curious to hear more about them.

Reply:
Happy to talk them through with you whenever